src/core/job.c

```diff
@@ -1048,14 +1052,15 @@ int job_finish_and_invalidate(Job *j, JobResult result, bool recursive, bool alr
          * this context. And JOB_FAILURE is already handled by the
          * unit itself. */
         if (IN_SET(result, JOB_TIMEOUT, JOB_DEPENDENCY)) {
-                log_struct(LOG_NOTICE,
-                           "JOB_TYPE=%s", job_type_to_string(t),
-                           "JOB_RESULT=%s", job_result_to_string(result),
-                           LOG_UNIT_ID(u),
-                           LOG_UNIT_MESSAGE(u, "Job %s/%s failed with result '%s'.",
-                                            u->id,
-                                            job_type_to_string(t),
-                                            job_result_to_string(result)));
+                if (!unit_log_level_test(u, LOG_NOTICE))
+                        log_struct(LOG_NOTICE,
+                                   "JOB_TYPE=%s", job_type_to_string(t),
+                                   "JOB_RESULT=%s", job_result_to_string(result),
+                                   LOG_UNIT_ID(u),
+                                   LOG_UNIT_MESSAGE(u, "Job %s/%s failed with result '%s'.",
+                                                    u->id,
+                                                    job_type_to_string(t),
+                                                    job_result_to_string(result)));
```

Comment:
@poettering This looks like it should be a `log_unit_struct` call as well; the reason it isn't is that, unlike all the other calls to `log_struct` that I modified, this one doesn't contain a `LOG_UNIT_INVOCATION_ID` field. I don't know if that was an oversight (and thus that this should be a `log_unit_struct` call like the others) or if there's a reason to exclude that field from this call; and if the latter, is this an acceptable solution, or should I remove `LOG_UNIT_INVOCATION_ID` from `log_unit_struct` and restore it to all the other call sites?

Reply:
the invocation ID only makes sense when we log about units that are currently active (or at least were just active). but a job can can be enqueued/failed long before a unit is started, i.e. long bevor the invocation ID is assigned to it... Hence, we don't add it here, becuse the job is not directly associated to a specific invocation of a unit, it's usually preparation for a state change, and that's quite often to start something which will then, but only then result in an invocatin ID allocation.
hence, I'd suggest LOG_UNIT_INVOCATION_ID() is something that should be passed manually to log_unit_strruct(), even if we'll pass it in 90% of the cases